### PR TITLE
parallel run by copy

### PR DESCRIPTION
**Pull request by lecourtoisflo**
linked to https://github.com/dynawo/dynawo/pull/1682

### CODE REVIEW

File: sources/Launcher/DYNMarginCalculationLauncher.cpp [hunk 1]
```diff
@@ -440,47 +450,45 @@ MarginCalculationLauncher::prepareEvents2Run(const task_t& requestedTask,
 }
 
 void
-MarginCalculationLauncher::launchScenario(const boost::shared_ptr<Scenario>& scenario, const std::string& baseJobsFile,
+MarginCalculationLauncher::launchScenario(const AnalysisContext& context, const boost::shared_ptr<Scenario>& scenario,
     const double variation, SimulationResult& result) {
-  if (nbThreads_ == 1)
-    std::cout << " Launch task :" << scenario->getId() << " dydFile =" << scenario->getDydFile() << std::endl;
+  std::stringstream ss;
```
Comment (rosiereflo) [inline sources/Launcher/DYNMarginCalculationLauncher.cpp:455]: Please keep the condition on number of threads

File: sources/Launcher/DYNMarginCalculationLauncher.cpp [hunk 2]
```diff
@@ -440,47 +450,45 @@ MarginCalculationLauncher::prepareEvents2Run(const task_t& requestedTask,
...
   result.setVariation(scenarioId.str());
-  boost::shared_ptr<DYN::Simulation> simulation = createAndInitSimulation(workingDir, job, params, result);
+  boost::shared_ptr<DYN::Simulation> simulation = createAndInitSimulation(workingDir, job, params, result, context);
 
   if (simulation)
     simulate(simulation, result);
-  if (nbThreads_ == 1)
-    std::cout << " Task :" << scenario->getId() << " status =" << getStatusAsString(result.getStatus()) << std::endl;
+
+  ss.str("");
```
Comment (rosiereflo) [inline sources/Launcher/DYNMarginCalculationLauncher.cpp:480]: same as above

File: sources/Launcher/DYNMarginCalculationLauncher.cpp [hunk 3]
```diff
@@ -535,15 +547,14 @@ MarginCalculationLauncher::findOrLaunchLoadIncrease(const boost::shared_ptr<Load
 void
 MarginCalculationLauncher::launchLoadIncrease(const boost::shared_ptr<LoadIncrease>& loadIncrease,
     const double variation, SimulationResult& result) {
-  if (nbThreads_ == 1)
-    std::cout << "Launch loadIncrease of " << variation << "%" <<std::endl;
+  std::stringstream ss;
```
Comment (rosiereflo) [inline sources/Launcher/DYNMarginCalculationLauncher.cpp:550]: Please keep the condition on number of threads

File: sources/Launcher/DYNMarginCalculationLauncher.h [hunk 4]
```diff
@@ -185,13 +185,14 @@ class MarginCalculationLauncher : public RobustnessAnalysisLauncher {
    * @brief launch the calculation of one scenario
    * Warning: must remain thread-safe!
    *
+   * @param context the analysis context to uyse
```
Comment (rosiereflo) [inline sources/Launcher/DYNMarginCalculationLauncher.h:188]: typo use

File: sources/Launcher/DYNMarginCalculationLauncher.h [hunk 5]
```diff
@@ -218,6 +219,13 @@ class MarginCalculationLauncher : public RobustnessAnalysisLauncher {
    */
   void initLog();
 
+  /**
+   * @brief Compute the scenario IIDM file corresponding to a variation
```
Comment (rosiereflo) [inline sources/Launcher/DYNMarginCalculationLauncher.h:223]: it generates the name of the corresponding iidm file

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: CPPLINT.cfg [hunk 6]
```diff
@@ -6,6 +6,7 @@ linelength=160
 # Filter checks
 filter=-legal/copyright
 filter=-build/include
+filter=-build/c++11
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply CPPLINT.cfg:9]: See remark on dynawo patch

File: sources/Launcher/DYNAnalysisContext.cpp [hunk 7]
```diff
@@ -0,0 +1,76 @@
+//
+// Copyright (c) 2021, RTE (http://www.rte-france.com)
+// See AUTHORS.txt
+// All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at http://mozilla.org/MPL/2.0/.
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source suite
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNAnalysisContext.cpp:10]: OK

File: sources/Launcher/DYNAnalysisContext.cpp [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+// of simulation tools for power systems.
+//
+
+/**
+ * @file  DYNAnalysisContext.cpp
+ *
+ * @brief Analysis context for data interface
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNAnalysisContext.cpp:17]: replaced by "Manager for data interface and job management during analysis launch"

File: sources/Launcher/DYNAnalysisContext.h [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+// Copyright (c) 2021, RTE (http://www.rte-france.com)
+// See AUTHORS.txt
+// All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at http://mozilla.org/MPL/2.0/.
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source suite
+// of simulation tools for power systems.
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNAnalysisContext.h:11]: OK

File: sources/Launcher/DYNAnalysisContext.h [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+namespace DYNAlgorithms {
+
+/**
+ * @brief Analysis context for data interface re-use
+ */
+class AnalysisContext {
+ public:
+  /**
+   * @brief Update analysis context
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNAnalysisContext.h:37]: OK

File: sources/Launcher/DYNAnalysisContext.h [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  void update(const std::string& workingDirectory, const std::string& jobFile, unsigned int nbVariants, const std::string& iidmFile = "");
+
+  /**
+   * @brief Update variant for current run
+   *
+   * The variant name will be the string representation of the integer
+   *
+   * @param variant the variant number
+   */
+  void updateCurrentRun(unsigned int variant);
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNAnalysisContext.h:53]: OK

File: sources/Launcher/DYNDataInterfaceContainer.cpp [hunk 12]
```diff
@@ -0,0 +1,96 @@
+//
+// Copyright (c) 2021, RTE (http://www.rte-france.com)
+// See AUTHORS.txt
+// All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at http://mozilla.org/MPL/2.0/.
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source suite
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNDataInterfaceContainer.cpp:10]: OK

File: sources/Launcher/DYNDataInterfaceContainer.cpp [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+// of simulation tools for power systems.
+//
+
+/**
+ * @file  DYNDataInterfaceContainer.cpp
+ *
+ * @brief Data interface container implementation
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNDataInterfaceContainer.cpp:17]: OK

File: sources/Launcher/DYNDataInterfaceContainer.cpp [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+  mutable std::mutex mutex_;
+};
+#else
+
+class DataInterfaceContainer::DataInterfaceContainerImpl {
+ public:
+  explicit DataInterfaceContainerImpl(const boost::shared_ptr<DYN::DataInterface>& data) : referenceDataInterface(data) {}
+
+  void initDataInterface() {
+    dataInterface = referenceDataInterface->clone();
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNDataInterfaceContainer.cpp:68]: it is usefull. This avoid the reference data interface to be modified by the future run, allowing the next runs to be safe

File: sources/Launcher/DYNDataInterfaceContainer.h [hunk 15]
```diff
@@ -0,0 +1,63 @@
+//
+// Copyright (c) 2021, RTE (http://www.rte-france.com)
+// See AUTHORS.txt
+// All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at http://mozilla.org/MPL/2.0/.
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source suite
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNDataInterfaceContainer.h:10]: OK

File: sources/Launcher/DYNRobustnessAnalysisLauncher.cpp [hunk 16]
```diff
@@ -39,12 +39,18 @@
 #include <JOBDynModelsEntry.h>
 #include <JOBDynModelsEntryFactory.h>
 #include <JOBModelerEntry.h>
+#include <JOBXmlImporter.h>
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNRobustnessAnalysisLauncher.cpp:42]: Artefact of developement. Will be removed

File: sources/Launcher/DYNRobustnessAnalysisLauncher.h [hunk 17]
```diff
@@ -173,6 +180,8 @@ class RobustnessAnalysisLauncher {
   int nbThreads_;  ///< number of threads to use
   boost::shared_ptr<multipleJobs::MultipleJobs> multipleJobs_;  ///< multipleJobs description tu use for the systematic analysis
 
+  AnalysisContext context_;  ///< basic analysis context, common to all
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNRobustnessAnalysisLauncher.h:183]: Used as base context.

File: sources/Launcher/DYNSystematicAnalysisLauncher.cpp [hunk 18]
```diff
@@ -76,33 +77,37 @@ SystematicAnalysisLauncher::launch() {
...
+    results_[i] = launchScenario(events[i]);
+  }
 }
 
 SimulationResult
-SystematicAnalysisLauncher::launchScenario(const boost::shared_ptr<Scenario>& scenario, const std::string& baseJobsFile) {
-  if (nbThreads_ == 1)
-    std::cout << " Launch scenario :" << scenario->getId() << " dydFile =" << scenario->getDydFile() << std::endl;
+SystematicAnalysisLauncher::launchScenario(const boost::shared_ptr<Scenario>& scenario) {
+  std::stringstream ss;
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNSystematicAnalysisLauncher.cpp:91]: What's the rationnal ?

File: sources/Launcher/DYNSystematicAnalysisLauncher.cpp [hunk 19]
```diff
@@ -76,33 +77,37 @@ SystematicAnalysisLauncher::launch() {
...
+  ss << " Launch scenario :" << scenario->getId() << " dydFile =" << scenario->getDydFile() << std::endl;
+  std::cout << ss.str();
 
   std::string workingDir  = createAbsolutePath(scenario->getId(), workingDirectory_);
-  job::XmlImporter importer;
-  boost::shared_ptr<job::JobsCollection> jobsCollection = importer.importFromFile(workingDirectory_ + "/" + baseJobsFile);
-  //  implicit : only one job per file
-  job::job_iterator itJobEntry = jobsCollection->begin();
-  boost::shared_ptr<job::JobEntry>& job = *itJobEntry;
+  boost::shared_ptr<job::JobEntry> job = boost::make_shared<job::JobEntry>(*context_.jobEntry());
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNSystematicAnalysisLauncher.cpp:96]: The job entry will be modified by the addition of the dyd file

File: sources/Launcher/test/TestDataInterfaceContainer.cpp [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+
+  std::thread thread2([&container, &dataInterface2](){
+      container.initDataInterface();
+      dataInterface2 = container.getDataInterface();
+  });
+
+  thread1.join();
+  thread2.join();
+
+  ASSERT_NE(dataInterface1, dataInterface);
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/test/TestDataInterfaceContainer.cpp:68]: Add a test for null pointer before.

File: sources/Launcher/DYNDataInterfaceContainer.cpp [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+#endif
+
+namespace DYNAlgorithms {
+
+#ifdef LANG_CXX11
+class DataInterfaceContainer::DataInterfaceContainerImpl {
+ public:
+  explicit DataInterfaceContainerImpl(const boost::shared_ptr<DYN::DataInterface>& data) : referenceDataInterface(data) {}
+
+  void initDataInterface() {
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNDataInterfaceContainer.cpp:38]: The idea is precisely to hide the mechanism by thread

File: sources/Launcher/DYNMarginCalculationLauncher.cpp [hunk 22]
```diff
@@ -630,5 +641,12 @@ MarginCalculationLauncher::createOutputs(std::map<std::string, std::string>& map
   }
 }
 
+std::string
+MarginCalculationLauncher::computeScenarioIIDMFile(double variation) const {
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNMarginCalculationLauncher.cpp:645]: OK

File: sources/Launcher/DYNMarginCalculationLauncher.cpp [hunk 23]
```diff
@@ -407,11 +412,16 @@ void MarginCalculationLauncher::findOrLaunchScenarios(const std::string& baseJob
...
+  std::string iidmFile = computeScenarioIIDMFile(newVariation);
+  contextsByIIDM_[iidmFile].update(workingDirectory_, baseJobsFile, events2Run.size());
+
 #pragma omp parallel for schedule(dynamic, 1)
   for (unsigned int i=0; i < events2Run.size(); i++) {
+    contextsByIIDM_.at(iidmFile).updateCurrentRun(i);
     double variation = events2Run[i].second;
     size_t eventIdx = events2Run[i].first;
-    launchScenario(events[eventIdx], baseJobsFile, variation, scenariosCache_[variation].getResult(eventIdx));
+    launchScenario(contextsByIIDM_[iidmFile], events[eventIdx], variation, scenariosCache_[variation].getResult(eventIdx));
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNMarginCalculationLauncher.cpp:424]: the map key is the IIDM path so it should be OK

File: sources/Launcher/DYNMarginCalculationLauncher.h (hunk 5 above)
Comment (lecourtoisflo) [reply sources/Launcher/DYNMarginCalculationLauncher.h:223]: OK

File: sources/Launcher/DYNSystematicAnalysisLauncher.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply sources/Launcher/DYNSystematicAnalysisLauncher.cpp:91]: yes, because we do not want to see messages mixed with each others on the standard output due to MT

File: sources/Launcher/DYNMarginCalculationLauncher.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply sources/Launcher/DYNMarginCalculationLauncher.cpp:424]: yes but iidmFile is not regenerated with the local variation value?

[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNMarginCalculationLauncher.cpp:424]: OK

File: sources/Launcher/DYNMarginCalculationLauncher.cpp (hunk 1 above)
Comment (lecourtoisflo) [reply sources/Launcher/DYNMarginCalculationLauncher.cpp:467]: OK

File: sources/Launcher/DYNMarginCalculationLauncher.cpp (hunk 2 above)
Comment (lecourtoisflo) [reply sources/Launcher/DYNMarginCalculationLauncher.cpp:492]: OK

File: sources/Launcher/DYNMarginCalculationLauncher.cpp (hunk 3 above)
Comment (lecourtoisflo) [reply sources/Launcher/DYNMarginCalculationLauncher.cpp:562]: OK

File: sources/Launcher/DYNSystematicAnalysisLauncher.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNSystematicAnalysisLauncher.cpp:91]: OK

File: sources/Launcher/DYNSystematicAnalysisLauncher.cpp [hunk 24]
```diff
@@ -76,33 +77,37 @@ SystematicAnalysisLauncher::launch() {
...
-  boost::shared_ptr<DYN::Simulation> simulation = createAndInitSimulation(workingDir, job, params, result);
+  boost::shared_ptr<DYN::Simulation> simulation = createAndInitSimulation(workingDir, job, params, result, context_);
 
   if (simulation) {
     simulate(simulation, result);
   }
-  if (nbThreads_ == 1)
-    std::cout << " scenario :" << scenario->getId() << " final status: " << getStatusAsString(result.getStatus()) << std::endl;
+
+  ss.str("");
```
[the remark this answers is not in the document]
Comment (lecourtoisflo) [reply sources/Launcher/DYNSystematicAnalysisLauncher.cpp:107]: OK
